lib/endpoint-proxy.js

```diff
@@ -521,11 +768,92 @@ class EndpointProxy {
     }
   }
 }
+class ServiceCaller {
+  constructor(client, serviceDeclaration, endpointDeclarations) {
+    if (client === void 0 || client === null) {
+      throw new Error('no database client provided');
+    }
+    if (serviceDeclaration === void 0 || serviceDeclaration === null) {
+      throw new Error('no service declaration provided');
+    } else if (!(typeof serviceDeclaration.endpointExtension === 'string' || (serviceDeclaration.endpointExtension instanceof String))) {
+      throw new Error('serviceDeclaration.endpointExtension must be a String');
+    }
+
+    if(endpointDeclarations === null || (Array.isArray(endpointDeclarations) && endpointDeclarations.length > 0)) {
+      throw new Error('endpointDeclarations cannot be null or empty.');
+    }
+
+    const db = client.getConnectionParams().database;
+    if (db !== void 0 && db !== null) {
+      throw new Error('connection for client cannot specify database');
+    }
+
+    this.$mlProxy = endpointDeclarations.reduce((proxy,  endpointDeclarations) =>
+    {
+      expandFunctionDeclaration(endpointDeclarations);
+      return proxy.withFunction(endpointDeclarations, serviceDeclaration.endpointExtension);
+    }, client.createProxy(client, serviceDeclaration)
+  );
+  }
+
+  call(functionName, args) {
+    return this.$mlProxy.execute(functionName, args, 0);
+  }
+}
+
+class EndpointCaller {
+  constructor(client, endpointDeclaration) {
+    if (client === void 0 || client === null) {
+      throw new Error('no database client provided');
+    }
+    if (endpointDeclaration === void 0 || endpointDeclaration === null) {
+      throw new Error('endpointDeclaration cannot be null.');
+    }
+    const endpoint = endpointDeclaration.endpoint;
+
+    if (endpoint === void 0 || endpoint === null) {
+      throw new Error('endpointDeclaration endpoint cannot be null.');
+    }
+    const regex = /^(.*[/])([^/]+)(.[^/.]+)$/;
+    const arrayValues = endpoint.match(regex);
```

Comment:
Would it be defensive to test that the array has at least 4 items (the [0] item is the entire match; the other items are the matching groups)?  

That way, input that doesn't match the regex wouldn't trigger an array access error on line 820?

Also, would endpointFields be a more semantic name and thus easier to understand than arrayValues (which is syntactic)?  

By the way, the regex was just a guess -- if you haven't done so, it would be a good idea to execute the match() expression in Query Console or any other JavaScript environment on some paths that should be expected to succeed or fail, as in:

- /dir/file.extension - should succeed
-  /file.extension - should succeed
- file.extension - should fail
- /dir/file - should fail
- /dir/.extension - should fail